lib/std/special/compiler_rt/emutls.zig

```diff
@@ -0,0 +1,356 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2015-2018 LLVM Compiler Infrastructure
+// Copyright (c) 2020 Sebastien Marie <semarie@online.fr>
+// This file is part of [zig](https://ziglang.org/), which is MIT licensed.
+// The MIT license requires this copyright notice to be included in all copies
+// and substantial portions of the software.
+// __emutls_get_address specific builtin
+//
+// derived work from LLVM Compiler Infrastructure - release 8.0 (MIT)
+// https://github.com/llvm-mirror/compiler-rt/blob/release_80/lib/builtins/emutls.c
+//
+
+const std = @import("std");
+const Allocator = std.mem.Allocator;
+
+comptime {
+    assert(std.builtin.link_libc);
+    assert(std.Thread.use_pthreads);
+}
+
+/// public entrypoint for generated code using EmulatedTLS
+pub fn __emutls_get_address(control: *emutls_control) callconv(.C) *c_void {
+    const allocator = std.heap.c_allocator; // XXX
+
+    return control.get_pointer(allocator) catch {
+        @panic("emutls_control.get_pointer");
```

Comment:
My personal choice here is to use `os.abort()`, trying to trigger a panic here is hopeless as you may end up calling into `__emutls_get_address` again and get a nice infinite loop.

Ditto for all the uses of `@panic`.

Reply:
Right: `@panic()` relies on a `threadvar`.